staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	//
+	// The language is as defined in
+	// https://kubernetes.io/docs/reference/using-api/cel/
+	// with several additions that are specific to device selectors.
+	//
+	// Attributes of a device are made available through a nested
+	// `device.attributes` map with the domain part of the attribute name
+	// as key in the outer map and the identifier as key in the inner
+	// map. All identifiers can be used in a field lookup:
+	//
+	//    device.attributes["dra.example.com"].driverVersion
+	//
+	// The type of each entry varies, depending on the attribute
+	// that is being looked up. The domain lookup returns an empty
+	// map if there is no attribute with that domain. However,
+	// unknown identifiers then trigger a runtime error.
+	//
+	// The `cel.bind` function is enabled and can be used to simplify
+	// expressions that access multiple attributes with the same domain:
+	//
+	//    cel.bind(dra, device.attributes["dra.example.com"], dra.someBool && dra.anotherBool)
+	//
+	// Capacities associated with a device are made available through a
+	// nested `device.capacities` map the same way as attributes.
+	//
+	// The `device.driverName` string variable can be used to check for a specific
+	// driver explicitly in a filter that is meant to work for devices from
+	// different vendors. It is provided by Kubernetes and matches the
+	// `driverName` from the ResourceSlice which provides the device.
+	//
+	// The CEL expression is applied to *all* available devices from any driver.
+	// The expression has to check for existence of an attribute when it is not
+	// certain that it is provided because runtime errors are not automatically
+	// treated as "don't select device". Instead, device selection fails completely
+	// and reports the error.
+	//
+	// Some more examples:
+	//
+	//    "memory" in device.capacities["dra.example.com"] && # Is the capacity available?
+	//       device.capacities["dra.example.com"].memory.isGreaterThan(quantity("1Gi")) # >= 1Gi
+	//
+	//    device.attributes["dra.example.com"].driverVersion.isGreaterThan(semver("1.0.0")) # >= v1.0.0, runtime error if not available
+	//
+	//    device.driverName == "dra.example.com" # any device from that driver
+	//
+	// +required
+	Expression string `json:"expression" protobuf:"bytes,1,name=expression"`
+}
+
+// Besides the request name slice, constraint must have exactly one field set.
+type Constraint struct {
+	// The constraint applies to devices in these requests. A single entry is okay
+	// and used when that request is for multiple devices.
+	//
+	// If empty, the constrain applies to all devices in the claim.
+	//
+	// +optional
+	// +listType=atomic
+	RequestNames []string `json:"requestNames,omitempty" protobuf:"bytes,1,opt,name=requestNames"`
+
+	// The devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "dra.example.com/numa" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
```

Comment:
How do we treat a device which does not have the named attribute?  Are we implicitly making this do `has()` before reading the value?  

What if 2 devices have the attribute but different types?

Reply:
> How do we treat a device which does not have the named attribute? 

Then it does not satisfy the constraint. We could allow "all devices must not have it", but I think the expectation is that this is a field which must be set.

> What if 2 devices have the attribute but different types

Then they don't match. It would be a bug in one driver, because the type is part of the definition of `dra.example.com/num`.

Reply:
Let's document the behavior, especially in the case of does-not-exist.

Reply:
IMHO that is part of "must have this attribute". I added more text in the example paragraph about this to make it clearer.